### ISSUE TITLE
Allow bnd-maven-plugin standard inheritable configuration in pom

### ISSUE THREAD

**Issue by amichair** [bndtools/bnd#952]
The bnd-maven-plugin announcement blog states that taking the configuration from an external bnd.bnd file in each project "rather than from a funky XML encoding of those instructions deep inside the POM" is an advantage of the new plugin.

I'd like to request that this be reconsidered, and that the plugin be made to also support a standard pom configuration section, with all the expected behavior as such.

If a user doesn't like poms or xmls, he is free to choose a build tool that doesn't use them. But as a Maven plugin maintainer, it makes much more sense to play along with the Maven conventions (especially since Maven touts conventions as its guiding philosophy), which means behaving like most other plugins, with the same form of funky XML encodings deep inside the POM.

This would mean you get all the Maven plugin behavior you'd expect intuitively by convention - configuration inheritance, variable interpolation, working with existing tools that support pom maintenance (change detection, syntax highlighting, whatever), etc.

As it stands, imagine a real-world project with a dozen or two bundles as submodules. Currently, instead of having the option to configure it once in a parent pom (with variables, etc.) and work on all modules, as is possible with the maven bundle plugin and all other Maven plugins, you'd now have to duplicate the bnd.bnd file many times, clutter the projects, and make sure they are all properly updated.

Just imagine how a Maven project would look if every plugin opted to use an external configuration file rather than configuring it within the pom... every module of every project would be littered with a couple dozen files in various syntaxes and forms... I suspect Maven would never have become popular if that were the case.

In short... please reconsider using the standard pom configuration section with all the built-in and conventional perks, and play along with the Maven spirit. Like it or not, it's a Maven plugin you're trying to provide, not an external build tool (well that too, but not in this sub-project :-) ). Using an external bnd file can remain optional to those who want it, and can be merged with the inline pom configuration similar to how manifest files are processed by the jar plugin and others.

**Comment by bjhargrave** [comment]
The master branch (for 3.0 of bnd-maven-plugin) has been updated to use maven project properties in the bnd file: https://github.com/bndtools/bnd/commit/530a664d9de45da89cba6a6f4ff581c41b91ea68. So any ${project.xxx} properties defined in POMs can be used in the bnd file. This means you can define things in the POM and just reference them in the bnd file.

Having two ways to do a thing (bnd file and maven configuration) is not always good. But supporting maven project properties in the bnd file will allow you to concentrate the definitions of things in the POM if you desire.

**Comment by bjhargrave** [comment]
> please reconsider using the standard pom configuration section with all the built-in and conventional perks

Also, every time we add a new property to bnd (which happens a lot), we would also need to make sure to add it to the XML configuration for bnd-maven-plugin. Keeping the bnd properties and the XML configuration in sync will be a loosing battle.

**Comment by amichair** [comment]
- Regarding interpolating properties in the bnd file - that's a good start :-) Any idea when 3.0 might be released?

- As for having two ways to do things - it's true that it's not always good, but in this case it seems like a great compromise. Would a request be considered to abandon the bnd files and implement only the pom configuration? I personally wouldn't mind that, but perhaps others would.

- If I understand the xml/properties sync concern above correctly, I don't think it's a problem. I was thinking of something like how the jar plugin (or plexus archive within?) lets you set manifestEntries to customize the manifest file - you can add arbitrary elements which are dynamically added as extra manifest name/value entries. So I'm sure it's possible to just take the xml elements and convert them dynamically to bnd properties (maybe just e.g. change to lowercase if necessary), and I don't think there will be any maintenance involved in adding properties with such a scheme.

- Another concern I have about the bnd file implementation is about inheritance/extendability of configuration. For example, in one project we have ~20 modules, each built into a bundle. The current bundle plugin configuration is in the parent, and they all inherit it automatically, other than having packaging type 'bundle' and a property specifying the bundle symbolic name (interpolated in parent configuration as expected). A couple of the bundles override/extend the configuration, e.g. for a WAB web bundle. How would this look with config files? Can a single config file be easily shared among all modules? And extended if necessary? If there's an elegant workaround that doesn't require duplicating the bnd config file 20 times, then it would really help in deciding whether to migrate to the new plugin. I haven't tried a full migration yet, so maybe there's something simple I've missed.

Thanks!

**Comment by rotty3000** [comment]
-include: ${parent.location}/parent.bnd

**Comment by bjhargrave** [comment]
> Regarding interpolating properties in the bnd file - that's a good start :-) Any idea when 3.0 might be released?

The plan is this summer.

> As for having two ways to do things - it's true that it's not always good, but in this case it seems like a great compromise. Would a request be considered to abandon the bnd files and implement only the pom configuration? I personally wouldn't mind that, but perhaps others would.

There is no way bnd will abandon bnd files :-) The whole point of the bnd-maven-plugin is to allow maven builds to create bundles using bnd and bnd uses bnd files.

The maven-bnd-plugin in Apache Felix uses xml configuration to drive bnd, so that choice is also available to you as well.

> If I understand the xml/properties sync concern above correctly, I don't think it's a problem. I was thinking of something like how the jar plugin (or plexus archive within?) lets you set manifestEntries to customize the manifest file - you can add arbitrary elements which are dynamically added as extra manifest name/value entries. So I'm sure it's possible to just take the xml elements and convert them dynamically to bnd properties (maybe just e.g. change to lowercase if necessary), and I don't think there will be any maintenance involved in adding properties with such a scheme.

The maven-bnd-plugin has xml configuation and it is a bit of a mess (IMO) in the mapping from xml elements to bnd properties. I don't see that we want to do that all over again here.

> Another concern I have about the bnd file implementation is about inheritance/extendability of configuration. For example, in one project we have ~20 modules, each built into a bundle. The current bundle plugin configuration is in the parent, and they all inherit it automatically, other than having packaging type 'bundle' and a property specifying the bundle symbolic name (interpolated in parent configuration as expected). A couple of the bundles override/extend the configuration, e.g. for a WAB web bundle. How would this look with config files? Can a single config file be easily shared among all modules? And extended if necessary? If there's an elegant workaround that doesn't require duplicating the bnd config file 20 times, then it would really help in deciding whether to migrate to the new plugin. I haven't tried a full migration yet, so maybe there's something simple I've missed.

As Ray mentions, bnd allow including of other bnd files. So you can include a parent bnd file (which can include a parent bnd file...). That plus using pom properties in the bnd files should prevent duplicating bnd information in every project.

**Comment by amichair** [comment]
Thanks for the responses.

The include looks like it would only work if it found the bnd config file on a fixed (relative) location on disk. In one project I'm working on, the bundles can be built independently, i.e. using the parent module that is deployed in the repo, not from a known path on disk. That's the more decoupled scenario I had in mind (sorry if I didn't make it clear).

I suppose you can close this issue as wontfix, since from the replies I understand that that is the case, and this plugin is not exactly what I hoped it would be. I guess my dream Maven bundle plugin would be doing everything the 'maven way' - using conventional pom configuration behavior like the Felix plugin, but with a more natural integration with the lifecycle and standard packaging like this new plugin (I really like this idea!) - i.e. the best of both worlds. But I now get that this bnd plugin has a different goal in mind, leaning more toward bnd users than toward Maven users, in some sense.

Thanks in any case for your work, it's really nice to see advances in this space :-)

**Comment by rotty3000** [comment]
There is also a null-safe `-include`. Just prefix the file name with `-`
thus if the file doesn't exist no harm done.

Raymond Augé

**Comment by pkriens** [comment]
I do agree that it makes sense to configure the maven way. I really would like Neil's input why the Felix's plugin instructions element is not used (and preferably with merging the instructions in any parents).

Neil?

**Comment by amichair** [comment]
I've found a good solution for my scenario for the time being - the Felix bundle plugin's manifest goal, bound to the process-classes phase, with supportedProjectType jar. So it integrates naturally with the Maven lifecycle similar to how the bnd plugin does it, playing along with all other standard plugins and without requiring bundle packaging, but with regular pom configuration so I can just configure it once in the parent pom and all the submodules just work (become bundles).

The only slight inconvenience, just like with the new bnd plugin, is that the jar plugin requires the external manifest file to be configured. The announcement blog mentions this needs to be done via the manifestFile configuration in the archive section, although the useDefaultManifestFile configuration (outside the archive section) is shorter and does the same thing afaik. The blog mentions that you're pushing to make this work without any configuration in the jar plugin itself, which is great. I assume if the generated manifest exists it will just pick it up automatically, or something like that, so it'll fix this for both plugins. Thanks for that :-)

**Comment by pkriens** [comment]
Since you found a solution we're going to close

**Comment by j3rem1e** [comment]
@pkriens as the solution here is to "not use this plugin", can the bnd-team reconsiders this issue and reopen it? ;-)

I'd like to use the bnd-maven-plugin, but without a proper way to inherit a configuration from a parent "corporate" pom, it's not possible.

**Comment by njbartlett** [comment]
@j3rem1e The bnd-maven-plugin does allow configuration to be inherited from the parent, so I'm not sure what else is missing. Is it just a matter of preferring to put everything in pom.xml?

**Comment by j3rem1e** [comment]
I am thinking about a "shared" parent pom : a pom which is not in the parent folder but shared in a maven repository. 

I have several projects, each project on his own git repository. Every project uses a shared parent pom to bring a common set of configuration (versions of libraries used, configurations of plugins like the compiler plugin .. and today the maven-bundle-plugin). This was the requirement of the OP of this issue too. 

You can see for example the "apache corporate pom", which provide basic configuration and information for all the apache project (http://svn.apache.org/viewvc/maven/pom/tags/apache-17/pom.xml?view=markup). I do the same, but we have a common configuration for bnd in this file (to be exact, a common configuration for the maven-bundle-plugin).

Today, bnd ignore this kind of shared pom, see this commit: https://github.com/bndtools/bnd/commit/c8883e4fc8ba41ea096a58d05443e4bbc96693e8#diff-e3072fb79083b871d9d9b44d62e81984

To be totally honest with you, I don't know another alternative than putting a common configuration in the pom.xml. Only this file can be uploaded to a maven repository and used like this.
